text/3720-labeled-match.md

```diff
@@ -0,0 +1,1452 @@
+
+Many other parser, decoder and other lowlevel crates will similarly benefit from the ergonomics of `loop match`.
+
+## Code generation
+
+State machines (parsers, interpreters, ect) can be written as a loop containing a match on the current state. The match picks the branch that belongs to the current state, some logic is performed, the state is updated, and eventually control flow jumps back to the top of the loop, branching to the next state.
+
+```rust
+loop {
+    match state {
+        A => {
+            // <perform work>
+
+            state = B;
+        }
+        B => {
+            // ...
+        }
+        // ...
+    }
+}
+```
+
+While this is a natural way to express a state machine, it is well-known that when translated to machine code in a straightforward way, this approach is inefficient on modern CPUs:
+
+- The match is an unpredictable branch, causing many branch misses. Reducing the number of branch misses is crucial for good performance on modern hardware.
+- The "loop + match" approach contains control flow paths (so, sequences of branches) that will never be taken in practice. More opimizations are possible if the actual possible paths are known more precicely (e.g. stack space can be reused if the value stored there will not be used in later states).
+
+By providing the compiler with more precise knowlege about what state transitions actually exists (i.e. what other states can follow a particular state), we get major performance improvements in practice. A proof of concept implementation of `loop match` shows considerable performance gains versus current recommended workarounds in real-world scenarios ([all results](https://gist.github.com/folkertdev/977183fb706b7693863bd7f358578292)):
+
+```
+Benchmark 3 (80 runs): /tmp/labeled-match-len rs-chunked 4 silesia-small.tar.gz
+  measurement          mean ± σ            min … max           outliers         delta
+  wall_time          62.6ms ±  555us    61.7ms … 66.1ms          2 ( 3%)        ⚡- 14.0% ±  1.4%
+  peak_rss           24.1MB ± 77.9KB    23.9MB … 24.1MB          0 ( 0%)          -  0.1% ±  0.1%
+  cpu_cycles          249M  ± 1.87M      248M  …  263M           5 ( 6%)        ⚡- 15.4% ±  1.3%
+  instructions        686M  ±  267       686M  …  686M           0 ( 0%)        ⚡- 24.9% ±  0.0%
+```
+
+So clearly, better code generation is possible, and not reliably achieved today.
+
+## Doesn't LLVM optimize this already?
+
+No.
+
+In some cases, the LLVM backend already achieves this optimal code generation using unconditional jumps, but the transformation is not guaranteed and fails for more complex inputs. Furthermore, LLVM is not the only rust codegen backend: it is likely that both `rustc_codegen_gcc` and `rustc_codegen_cranelift` will see more and more use. Hence we should be sceptical of relying on LLVM to achieve good codegen, and prefer performing optimization for all backends on the rustc MIR representation.
```

Comment:
I am simply a Rust user, but I have read over this and still don't see why new syntax and semantics are being added to the language to let the user tell LLVM what is going on instead of just making LLVM smarter to be able to pick up on this pattern or optimise this pattern in MIR (or a lower level version of MIR).

Overall this is giving me the vibes of `register` in C, in N years LLVM or whatever codegen backend could pick up on this pattern and make it entirely syntax cruft.

Reply:
There are significant differences to `register` and this is similar to tail calls (#3407).  What happens here under the hood is an optimization that is only possible in some cases, and even if the optimizer was perfect at detecting the pattern, a simple code change by the default could invalidate it.

This means that this code is quite fragile.  You could accidentally land a change that makes the optimization impossible and it would not be found in code review.  Adding annotations not just helps the optimizer, but also the compiler and the user.

Reply:
Got it, maybe add "Adds a level of guarantee that code generation does not regress over time" to the bullet points in the summary?